presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java

```diff
@@ -36,15 +39,22 @@
     public void configure(Binder binder)
     {
         configBinder(binder).bindConfig(CachingHiveMetastoreConfig.class);
+        newOptionalBinder(binder, HiveMetastoreDecorator.class);
         newExporter(binder).export(HiveMetastore.class)
                 .as(generator -> generator.generatedNameOf(CachingHiveMetastore.class));
     }
 
     @Provides
     @Singleton
-    public HiveMetastore createCachingHiveMetastore(@ForCachingHiveMetastore HiveMetastore delegate, @ForCachingHiveMetastore Executor executor, CachingHiveMetastoreConfig config)
+    public HiveMetastore createCachingHiveMetastore(
+            @ForCachingHiveMetastore HiveMetastore delegate,
+            @ForCachingHiveMetastore Executor executor,
+            CachingHiveMetastoreConfig config,
+            Optional<HiveMetastoreDecorator> hiveMetastoreDecorator)
     {
-        return cachingHiveMetastore(delegate, executor, config);
+        HiveMetastore decoratedDelegate = hiveMetastoreDecorator.map(decorator -> decorator.decorate(delegate))
+                .orElse(delegate);
+        return cachingHiveMetastore(decoratedDelegate, executor, config);
```

Comment:
Why its specific to caching?

Reply:
It is not specific to caching. Caching module is the one that provides actual `HiveMetastore` implementation.

Reply:
But why not to decorate caching metastore instead? Who provides ForCachingHiveMetastore? I think it should he decorated there.
Putting it here is odd

Reply:
That’s a good point. Do we want to decorate physical or logical calls? We should probably document this in the interface.

Reply:
> Who provides ForCachingHiveMetastore?

Every metastore module like FileHiveMetastoreModule etc.

> I think it should he decorated there.

Then I would need to decorate metastore in 4 places. Also possibly there might be Hive metastore that are provided externally, then I might not be able to do it there.

> Do we want to decorate physical or logical calls? 

I care only about physical calls. 

> We should probably document this in the interface.

Do you mean javadoc?

I think the best design would be to provide a new guice module between metastores modules and caching module, but I considered this as an overkill.

Reply:
> Every metastore module like FileHiveMetastoreModule etc.

I think another module and annotation (e.g `@ForHiveMetastoreDecorator`) would be cleaner solution

Reply:
Do you mean to using `@ForHiveMetastoreDecorator` instead of current `@ForHiveMetastoreCaching`?  And then module with decoration would produce bean with `@ForHiveMetastoreCaching`? 

I somehow think that this kind of a contract that every metastore module produce `@ForHiveMetastoreCaching`. I didn't want to change that.

Instead maybe we should not annotate beans with information where they should be consumed but with information what do they provide. So I propose to have:
 - `@BaseHiveMetastore`
 - `@DecoratedHiveMetastore`
 - `@CachingHiveMetastore`
Then last module would expose `@CachingHiveMetastore` as the `HiveMetastore`. 
WDYT?

Reply:
`@CachingHiveMetastore` would not be necessarily caching. I thin using `@ForXXX` abstracts actual providing implementation